[torch.fx] Pass placeholders through delegate too

dzhulgakov opened the pull request:
It's useful if we add additional attributed to nodes in the graph - it's easier to set the attribute on all nodes, even if the value would happen to be None.

jamesr66a reviewed (approved):
LGTM

jerryzh168 commented on torch/fx/graph.py:
```diff
@@ -117,12 +117,6 @@ def _name(self, target: Target) -> str:
         i = self._used_names[op] = self._used_names[op] + 1
         return f'{op}_{i}'
 
-    def get_param(self, name: str) -> Node:
```
can we keep these helper functions? or should we always create node through delegate now?

dzhulgakov replied:
Well, it's then easy to mix up delegate vs regular creation. Also, why placeholder and get_param but not others like call_module?

jerryzh168 replied:
> Well, it's then easy to mix up delegate vs regular creation.

so is delegate the suggested API for creating nodes?

> Also, why placeholder and get_param but not others like call_module?

I was planning to add helper functions for other node types as well.

jamesr66a replied:
Node creation should always go through a `delegate` and the `delegate` should be specifiable by the caller of the pass. Otherwise, we will lose composability, for example if I want to run a pipeline of passes on a Module but I want custom leaf modules

I might go so far as to say we should make `create_node` private, or move it into Delegate itself to make Graph just a dumb container for nodes

jerryzh168 replied:
always enforcing people to use delegate sounds good, are you planning to do it? I still have some prs on quantization, feel free to change all the other call sites of create_node first and I can have a PR to make create_node private and change quantization call sites.

jamesr66a commented on torch/fx/symbolic_trace.py:
```diff
@@ -55,6 +55,12 @@ def create_node(self, kind : str, target : Union[str, Callable],
+
+    def get_param(self, target):
+        return self.create_node('get_param', target, None, None)
+
```
You would describe when you would call these methods, the arguments and types, etc. They can be docstrings if you want

Commit:
[torch.fx] Pass placeholders through delegate too (#43432)

Summary:
It's useful if we add additional attributed to nodes in the graph - it's easier to set the attribute on all nodes, even if the value would happen to be None.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/43432

Reviewed By: jamesr66a

Differential Revision: D23276433

Pulled By: dzhulgakov

fbshipit-source-id: 11b3f452e7fe859c5d6c861ff42cd326dd3b66c0
